src/spec.rs

```diff
@@ -674,25 +524,234 @@ impl Zip64CentralDirectoryEnd {
+        writer.write_all(&comment)?;
+        Ok(())
+    }
+}
+
+pub(crate) struct CentralDirectoryEndInfo {
+    pub eocd: (Zip32CentralDirectoryEnd, u64),
+    pub eocd64: Option<(Zip64CentralDirectoryEnd, u64)>,
+
+    pub archive_offset: u64,
+}
+
+/// Finds the EOCD and possibly the EOCD64 block and determines the archive offset.
+///
+/// In the best case scenario (no prepended junk), this function will not backtrack
+/// in the reader.
+pub fn find_central_directory<R: Read + Seek>(
+    reader: &mut R,
+    archive_offset: ArchiveOffset,
+) -> ZipResult<CentralDirectoryEndInfo> {
+    const EOCD_SIG_BYTES: [u8; mem::size_of::<Magic>()] =
+        Magic::CENTRAL_DIRECTORY_END_SIGNATURE.to_le_bytes();
+
+    const EOCD64_SIG_BYTES: [u8; mem::size_of::<Magic>()] =
+        Magic::ZIP64_CENTRAL_DIRECTORY_END_SIGNATURE.to_le_bytes();
+
+    const CDFH_SIG_BYTES: [u8; mem::size_of::<Magic>()] =
+        Magic::CENTRAL_DIRECTORY_HEADER_SIGNATURE.to_le_bytes();
+
+    // Probe the file size
+    let file_length = reader.seek(io::SeekFrom::End(0))?;
+
+    // Instantiate the mandatory finder
+    let mut eocd_finder = MagicFinder::new(&EOCD_SIG_BYTES, (0, file_length));
+    let mut subfinder: Option<OptimisticMagicFinder<'static>> = None;
+
+    // Keep the last errors for cases of improper EOCD instances.
+    let mut parsing_error = None;
+
+    while let Some(eocd_offset) = eocd_finder.next_back(reader)? {
+        // Attempt to parse the EOCD block
+        let eocd = match Zip32CentralDirectoryEnd::parse(reader) {
+            Ok(eocd) => eocd,
+            Err(e) => {
+                if parsing_error.is_none() {
+                    parsing_error = Some(e);
+                }
+                continue;
+            }
+        };
+
+        // ! Relaxed (inequality) due to garbage-after-comment Python files
+        // Consistency check: the EOCD comment must terminate before the end of file
+        if eocd.zip_file_comment.len() as u64 + eocd_offset + 22 > file_length {
+            parsing_error = Some(ZipError::InvalidArchive("Invalid EOCD comment length"));
+            continue;
+        }
+
+        // Branch out for zip32
+        if !eocd.is_zip64() {
+            let relative_cd_offset = eocd.central_directory_offset as u64;
+
+            // If the archive is empty, there is nothing more to be checked, the archive is correct.
+            if eocd.number_of_files == 0 {
+                return Ok(CentralDirectoryEndInfo {
+                    eocd: (eocd, eocd_offset),
+                    eocd64: None,
+                    archive_offset: eocd_offset - relative_cd_offset,
+                });
+            }
+
+            // Consistency check: the CD relative offset cannot be after the EOCD
+            if relative_cd_offset >= eocd_offset {
+                parsing_error = Some(ZipError::InvalidArchive("Invalid CDFH offset in EOCD"));
+                continue;
+            }
+
+            // Attempt to find the first CDFH
```

Comment:
Don't you mean the last one?

Reply:
Yeah, this might actually be an error in the algorithm, because what I should be looking for is the first file (to correctly determine the archive offset). No idea how this gets past the tests, I will look into this.

Reply:
Yes, this is incorrect, the reason the test pass is that the initial guess is always right. I will add a test case where this is not true.

Reply:
I ended up implementing both directions for the MagicFinder and implemented the correct search direction. I think the only thing remaining is the may_be_zip64 logic for edge case zip32s.